File: packages/core/src/scanner/ScannerConfig.ts
import { readFileSync, existsSync } from 'node:fs';
import { parse as parseYaml } from 'yaml';

/**
 * Scanner configuration options
 */
export interface ScannerConfig {
  /** Allow remote LLM providers (openai, anthropic, gemini) */
  allowRemoteProviders: boolean;
  /** List of allowed provider names */
  allowedProviders?: string[];
  /** Preferred model names */
  modelNames?: string[];
  /** OpenAI API key */
  openaiApiKey?: string;
  /** Anthropic API key */
  anthropicApiKey?: string;
  /** Google Gemini API key */
  geminiApiKey?: string;
  /** Ollama endpoint URL */
  ollamaEndpoint?: string;
  /** Maximum retries for API calls */
  retryMaxRetries?: number;
  /** Delay between retries in ms */
  retryDelayMs?: number;
  /** Request timeout in ms */
  timeoutMs?: number;
  /** File extensions to include */
  includeExtensions?: string[];
  /** File extensions to exclude */
  excludeExtensions?: string[];
  /** Glob patterns to include */
  includeGlobs?: string[];
  /** Glob patterns to exclude */
  excludeGlobs?: string[];
  /** Maximum file size in bytes */
  maxFileBytes?: number;
  /** Patterns to exclude */
  excludePatterns?: string[];
  /** Patterns that override exclusions */
  includeOverride?: string[];
  /** Exclude documentation files */
  excludeDocs?: boolean;
  /** Exclude test files */
  excludeTests?: boolean;
  /** Enable LLM review of findings */
  llmReview?: boolean;
  /** Enable adaptive chunking based on token limits */
  adaptiveChunking?: boolean;
  /** Maximum prompt tokens */
  maxPromptTokens?: number;
  /** Reserve tokens for output */
  reserveOutputTokens?: number;
  /** Enable heuristic gating */
  gatingEnabled?: boolean;
  /** Shadow sample rate for gated chunks */
  shadowSampleRate?: number;
}

/**
 * Default configuration values
 */
export const ScannerConfigDefaults = {
  DEFAULT_MAX_FILE_BYTES: 5 * 1024 * 1024,
  DEFAULT_MAX_PROMPT_TOKENS: 6000,
  DEFAULT_RESERVE_OUTPUT_TOKENS: 1024,
  DEFAULT_OLLAMA_ENDPOINT: 'http://localhost:11434',
  GPT5_MAX_PROMPT_TOKENS: 200000,
  GPT5_RESERVE_OUTPUT_TOKENS: 8192,
  DEFAULT_RETRY_MAX_RETRIES: 3,
  DEFAULT_RETRY_DELAY_MS: 1000,
  DEFAULT_TIMEOUT_MS: 60000,
} as const;

/**
 * Load scanner configuration from a file path
 */
export function loadScannerConfig(path?: string): ScannerConfig {
  let base: ScannerConfig = { allowRemoteProviders: false };

  if (path && existsSync(path)) {
    const content = readFileSync(path, 'utf-8');
    if (path.endsWith('.json')) {
      base = JSON.parse(content) as ScannerConfig;
    } else {
      base = parseYaml(content) as ScannerConfig;
    }
  }

  return applyEnvOverrides(base);
}

/**
 * Apply environment variable overrides to configuration
 */
function applyEnvOverrides(config: ScannerConfig): ScannerConfig {
  const cfg = { ...config };

  cfg.allowRemoteProviders = envBool(
    'ALLOW_REMOTE_PROVIDERS',
    cfg.allowRemoteProviders
  );

  const allowedProviders = envList('ALLOWED_PROVIDERS');
  if (allowedProviders) cfg.allowedProviders = allowedProviders;

  const modelNames = envList('MODEL_NAMES');
  if (modelNames) cfg.modelNames = modelNames;

  const openaiKey = envString('OPENAI_API_KEY');
  if (openaiKey) cfg.openaiApiKey = openaiKey;

  const anthropicKey = envString('ANTHROPIC_API_KEY');
  if (anthropicKey) cfg.anthropicApiKey = anthropicKey;

  const geminiKey =
    envString('GEMINI_API_KEY') ??
    envString('GOOGLE_GENERATIVE_AI_API_KEY') ??
    envString('GOOGLE_API_KEY');
  if (geminiKey) cfg.geminiApiKey = geminiKey;

  const ollamaEndpoint = envString('OLLAMA_ENDPOINT');
  if (ollamaEndpoint) cfg.ollamaEndpoint = ollamaEndpoint;

  const retryMaxRetries = envInt('RETRY_MAX_RETRIES');
  if (retryMaxRetries !== undefined) cfg.retryMaxRetries = retryMaxRetries;

  const retryDelayMs = envInt('RETRY_DELAY_MS');
  if (retryDelayMs !== undefined) cfg.retryDelayMs = retryDelayMs;

  const timeoutMs = envInt('TIMEOUT_MS');
  if (timeoutMs !== undefined) cfg.timeoutMs = timeoutMs;

  const includeExtensions = envList('INCLUDE_EXTENSIONS');
  if (includeExtensions) cfg.includeExtensions = includeExtensions;

  const excludeExtensions = envList('EXCLUDE_EXTENSIONS');
  if (excludeExtensions) cfg.excludeExtensions = excludeExtensions;

  const includeGlobs = envList('INCLUDE_GLOBS');
  if (includeGlobs) cfg.includeGlobs = includeGlobs;

  const excludeGlobs = envList('EXCLUDE_GLOBS');
  if (excludeGlobs) cfg.excludeGlobs = excludeGlobs;

  const maxFileBytes = envInt('MAX_FILE_BYTES');
  if (maxFileBytes !== undefined) cfg.maxFileBytes = maxFileBytes;

  const excludePatterns = envList('SCANNER_EXCLUDE_PATTERNS');
  if (excludePatterns) cfg.excludePatterns = excludePatterns;

  const includeOverride = envList('SCANNER_INCLUDE_OVERRIDE');
  if (includeOverride) cfg.includeOverride = includeOverride;

  if ('SCANNER_EXCLUDE_DOCS' in process.env) {
    cfg.excludeDocs = envBool('SCANNER_EXCLUDE_DOCS', true);
  }

  if ('SCANNER_EXCLUDE_TESTS' in process.env) {
    cfg.excludeTests = envBool('SCANNER_EXCLUDE_TESTS', true);
  }

  if ('LLM_REVIEW' in process.env) {
    cfg.llmReview = envBool('LLM_REVIEW', false);
  }

  if ('ADAPTIVE_CHUNKING' in process.env) {
    cfg.adaptiveChunking = envBool('ADAPTIVE_CHUNKING', true);
  }

  const maxPromptTokens = envInt('MAX_PROMPT_TOKENS');
  if (maxPromptTokens !== undefined) cfg.maxPromptTokens = maxPromptTokens;

  const reserveOutputTokens = envInt('RESERVE_OUTPUT_TOKENS');
  if (reserveOutputTokens !== undefined)
    cfg.reserveOutputTokens = reserveOutputTokens;

  if ('GATING_ENABLED' in process.env) {
    cfg.gatingEnabled = envBool('GATING_ENABLED', true);
  }

  const shadowSampleRate = envFloat('SHADOW_SAMPLE_RATE');
  if (shadowSampleRate !== undefined) cfg.shadowSampleRate = shadowSampleRate;

  // Apply defaults
  cfg.maxFileBytes ??= ScannerConfigDefaults.DEFAULT_MAX_FILE_BYTES;
  cfg.adaptiveChunking ??= true;
  cfg.maxPromptTokens ??= ScannerConfigDefaults.DEFAULT_MAX_PROMPT_TOKENS;
  cfg.reserveOutputTokens ??=
    ScannerConfigDefaults.DEFAULT_RESERVE_OUTPUT_TOKENS;
  cfg.gatingEnabled ??= true;
  cfg.shadowSampleRate ??= 0.0;

  return cfg;
}

function envString(key: string): string | undefined {
  const value = process.env[key];
  return value && value.trim() ? value.trim() : undefined;
}

function envList(key: string): string[] | undefined {
  const raw = envString(key);
  if (!raw) return undefined;
  const list = raw
    .split(',')
    .map((s) => s.trim())
    .filter((s) => s);
  return list.length > 0 ? list : undefined;
}

function envBool(key: string, defaultValue: boolean): boolean {
  const value = process.env[key]?.trim().toLowerCase();
  if (value === '1' || value === 'true' || value === 'yes') return true;
  if (value === '0' || value === 'false' || value === 'no') return false;
  return defaultValue;
}

function envInt(key: string): number | undefined {
  const value = process.env[key];
  if (!value) return undefined;
  const parsed = parseInt(value, 10);
  return isNaN(parsed) ? undefined : parsed;
}

function envFloat(key: string): number | undefined {
  const value = process.env[key];
  if (!value) return undefined;
  const parsed = parseFloat(value);
  return isNaN(parsed) ? undefined : parsed;
}

File: src/main/kotlin/ai/astha/scantheplanet/idea/scanner/ScannerConfig.kt
package ai.astha.scantheplanet.idea.scanner

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.module.kotlin.KotlinModule
import com.fasterxml.jackson.module.kotlin.readValue
import java.nio.file.Path

data class ScannerConfig(
    val allowRemoteProviders: Boolean = false,
    val allowedProviders: List<String>? = null,
    val modelNames: List<String>? = null,
    val openaiApiKey: String? = null,
    val anthropicApiKey: String? = null,
    val geminiApiKey: String? = null,
    val ollamaEndpoint: String? = null,
    val ollamaLoggingEnabled: Boolean = false,
    val ollamaLogPath: String? = null,
    val retryMaxRetries: Int? = null,
    val retryDelayMs: Long? = null,
    val timeoutMs: Long? = null,
    val includeExtensions: List<String>? = null,
    val excludeExtensions: List<String>? = null,
    val includeGlobs: List<String>? = null,
    val excludeGlobs: List<String>? = null,
    val sourceCodeOnly: Boolean? = null,
    val maxFileBytes: Long? = null,
    val excludePatterns: List<String>? = null,
    val includeOverride: List<String>? = null,
    val excludeDocs: Boolean? = null,
    val excludeTests: Boolean? = null,
    val llmReview: Boolean? = null,
    val adaptiveChunking: Boolean? = null,
    val maxPromptTokens: Int? = null,
    val reserveOutputTokens: Int? = null,
    val gatingEnabled: Boolean? = null,
    val shadowSampleRate: Double? = null,
    val chunkParallelism: Int? = null,
    val chunkParallelismMax: Int? = null
)

class ScannerConfigLoader {
    private val yamlMapper = ObjectMapper(YAMLFactory()).registerModule(KotlinModule.Builder().build())
    private val jsonMapper = ObjectMapper().registerModule(KotlinModule.Builder().build())

    fun load(path: Path?): ScannerConfig {
        val base = if (path != null) {
            val content = path.toFile().readText(Charsets.UTF_8)
            if (path.toString().endsWith(".json")) {
                jsonMapper.readValue<ScannerConfig>(content)
            } else {
                yamlMapper.readValue<ScannerConfig>(content)
            }
        } else {
            ScannerConfig()
        }
        return applyEnvOverrides(base)
    }

    private fun applyEnvOverrides(config: ScannerConfig): ScannerConfig {
        var cfg = config
        val allowRemoteProviders = envBool("ALLOW_REMOTE_PROVIDERS", cfg.allowRemoteProviders)
        cfg = cfg.copy(allowRemoteProviders = allowRemoteProviders)

        envList("ALLOWED_PROVIDERS")?.let { cfg = cfg.copy(allowedProviders = it) }
        envList("MODEL_NAMES")?.let { cfg = cfg.copy(modelNames = it) }
        envString("OPENAI_API_KEY")?.let { cfg = cfg.copy(openaiApiKey = it) }
        envString("ANTHROPIC_API_KEY")?.let { cfg = cfg.copy(anthropicApiKey = it) }

        val gemini = envString("GEMINI_API_KEY")
            ?: envString("GOOGLE_GENERATIVE_AI_API_KEY")
            ?: envString("GOOGLE_API_KEY")
        if (!gemini.isNullOrBlank()) {
            cfg = cfg.copy(geminiApiKey = gemini)
        }
        envString("OLLAMA_ENDPOINT")?.let { cfg = cfg.copy(ollamaEndpoint = it) }
        if (System.getenv().containsKey("OLLAMA_LOGGING_ENABLED")) {
            cfg = cfg.copy(ollamaLoggingEnabled = envBool("OLLAMA_LOGGING_ENABLED", false))
        }
        envString("OLLAMA_LOG_PATH")?.let { cfg = cfg.copy(ollamaLogPath = it) }

        envInt("RETRY_MAX_RETRIES")?.let { cfg = cfg.copy(retryMaxRetries = it) }
        envLong("RETRY_DELAY_MS")?.let { cfg = cfg.copy(retryDelayMs = it) }
        envLong("TIMEOUT_MS")?.let { cfg = cfg.copy(timeoutMs = it) }
        envList("INCLUDE_EXTENSIONS")?.let { cfg = cfg.copy(includeExtensions = it) }
        envList("EXCLUDE_EXTENSIONS")?.let { cfg = cfg.copy(excludeExtensions = it) }
        envList("INCLUDE_GLOBS")?.let { cfg = cfg.copy(includeGlobs = it) }
        envList("EXCLUDE_GLOBS")?.let { cfg = cfg.copy(excludeGlobs = it) }
        if (System.getenv().containsKey("SCANNER_SOURCE_CODE_ONLY")) {
            cfg = cfg.copy(sourceCodeOnly = envBool("SCANNER_SOURCE_CODE_ONLY", false))
        }
        envLong("MAX_FILE_BYTES")?.let { cfg = cfg.copy(maxFileBytes = it) }
        envList("SCANNER_EXCLUDE_PATTERNS")?.let { cfg = cfg.copy(excludePatterns = it) }
        envList("SCANNER_INCLUDE_OVERRIDE")?.let { cfg = cfg.copy(includeOverride = it) }
        if (System.getenv().containsKey("SCANNER_EXCLUDE_DOCS")) {
            cfg = cfg.copy(excludeDocs = envBool("SCANNER_EXCLUDE_DOCS", true))
        }
        if (System.getenv().containsKey("SCANNER_EXCLUDE_TESTS")) {
            cfg = cfg.copy(excludeTests = envBool("SCANNER_EXCLUDE_TESTS", true))
        }
        if (System.getenv().containsKey("LLM_REVIEW")) {
            cfg = cfg.copy(llmReview = envBool("LLM_REVIEW", false))
        }
        if (System.getenv().containsKey("ADAPTIVE_CHUNKING")) {
            cfg = cfg.copy(adaptiveChunking = envBool("ADAPTIVE_CHUNKING", true))
        }
        envInt("MAX_PROMPT_TOKENS")?.let { cfg = cfg.copy(maxPromptTokens = it) }
        envInt("RESERVE_OUTPUT_TOKENS")?.let { cfg = cfg.copy(reserveOutputTokens = it) }
        if (System.getenv().containsKey("GATING_ENABLED")) {
            cfg = cfg.copy(gatingEnabled = envBool("GATING_ENABLED", true))
        }
        envDouble("SHADOW_SAMPLE_RATE")?.let { cfg = cfg.copy(shadowSampleRate = it) }
        envInt("SCANNER_CHUNK_PARALLELISM")?.let { cfg = cfg.copy(chunkParallelism = it) }
        envInt("SCANNER_CHUNK_PARALLELISM_MAX")?.let { cfg = cfg.copy(chunkParallelismMax = it) }

        val maxFileBytes = cfg.maxFileBytes ?: ScannerConfigDefaults.DEFAULT_MAX_FILE_BYTES
        val adaptiveChunking = cfg.adaptiveChunking ?: true
        val maxPromptTokens = cfg.maxPromptTokens ?: ScannerConfigDefaults.DEFAULT_MAX_PROMPT_TOKENS
        val reserveOutputTokens = cfg.reserveOutputTokens ?: ScannerConfigDefaults.DEFAULT_RESERVE_OUTPUT_TOKENS
        val gatingEnabled = cfg.gatingEnabled ?: true
        val shadowSampleRate = cfg.shadowSampleRate ?: 0.0
        val chunkParallelism = cfg.chunkParallelism
        val chunkParallelismMax = cfg.chunkParallelismMax
        cfg = cfg.copy(
            maxFileBytes = maxFileBytes,
            adaptiveChunking = adaptiveChunking,
            maxPromptTokens = maxPromptTokens,
            reserveOutputTokens = reserveOutputTokens,
            gatingEnabled = gatingEnabled,
            shadowSampleRate = shadowSampleRate,
            chunkParallelism = chunkParallelism,
            chunkParallelismMax = chunkParallelismMax
        )
        return cfg
    }

    private fun envString(key: String): String? = System.getenv(key)?.takeIf { it.isNotBlank() }

    private fun envList(key: String): List<String>? {
        val raw = envString(key) ?: return null
        val list = raw.split(',').map { it.trim() }.filter { it.isNotEmpty() }
        return if (list.isEmpty()) null else list
    }

    private fun envBool(key: String, default: Boolean): Boolean {
        return when (System.getenv(key)?.trim()?.lowercase()) {
            "1", "true", "yes" -> true
            "0", "false", "no" -> false
            null -> default
            else -> default
        }
    }

    private fun envInt(key: String): Int? = System.getenv(key)?.toIntOrNull()

    private fun envLong(key: String): Long? = System.getenv(key)?.toLongOrNull()

    private fun envDouble(key: String): Double? = System.getenv(key)?.toDoubleOrNull()

}

object ScannerConfigDefaults {
    const val DEFAULT_MAX_FILE_BYTES: Long = 5L * 1024 * 1024
    const val DEFAULT_MAX_PROMPT_TOKENS: Int = 6000
    const val DEFAULT_RESERVE_OUTPUT_TOKENS: Int = 1024
    const val DEFAULT_OLLAMA_ENDPOINT: String = "http://localhost:11434"
    const val GPT5_MAX_PROMPT_TOKENS: Int = 200000
    const val GPT5_RESERVE_OUTPUT_TOKENS: Int = 8192
}
